torch/distributed/_checkpoint/metadata.py

```diff
@@ -0,0 +1,47 @@
+from typing import Any, Dict, List, Union, Optional
+import torch
+from dataclasses import dataclass
+import pickle
+from torch.distributed.shard.sharded_tensor import ShardedTensor, ShardedTensorMetadata, ShardMetadata
+
+TENSOR_TYPE = Union[torch.Tensor, ShardedTensor]
+
+@dataclass
+class StorageMetadata:
+    shard_metadata: Optional[ShardMetadata]
+    # Unique identifier for this particular entity (Tensor or Shard of ShardedTensor)
+    storage_key: str
+    length: int
+    offset: int
+
+# Metadata for each param.
+@dataclass
+class ExtendedTensorMetadata:
+    # Details of Tensor/ShardedTensor (dtype, shape, sharding config etc.)
+    tensor_metadata: ShardedTensorMetadata
+    storage_metadata: List[StorageMetadata]
+
+@dataclass
+class Metadata:
+    # Metadata for the state dict.
+    state_dict_metadata: Dict[str, ExtendedTensorMetadata]
+
+    def __getstate__(self) -> bytes:
+        serialized = pickle.dumps(self.state_dict_metadata)
+        return serialized
+
+    def __setstate__(self, state: bytes):
+        # TODO, use pickle for this quick hack, must replace it with something else e.g. flatbuffer
+        # before serious use case,
+        self.state_dict_metadata = pickle.loads(state)
+
+@dataclass
+class ReadWriteRequest:
```

Comment:
This is not actually `metadata`? Probably belongs in the StorageReader/Writer classes. Also, this will probably have to move to C++ given that we want to replace `torch.Tensor` with a buffer.